### ISSUE TITLE
Tolerance parameter seems to have limited effect

### ISSUE THREAD

**Issue by 10110111** [yixuan/spectra#39]
I'm trying to calculate first 3 eigenpairs of a test 2500x2500 matrix, and Spectra appears to give quite low precision. See [this code](https://pastebin.mozilla.org/9077781), accompanied by the [reference eigenvectors file](https://pastebin.com/raw/8jvpzs7t), which you'll pass to the program. On my system I get the following output:
```
Eigenvalues found in 5424 iterations, 13820 operations:
0.493480219899482
0.493480218260062
0.197392088025424
Difference from reference eigenvalues:
-1.54972090715688e-10
-1.79439174630147e-09
3.65107943878229e-12
Standard deviations of H*v-lambda*v for computed eigenvectors:
5.92741914152269e-11
1.57294241003771e-10
2.39031776387999e-13
Standard deviations of H*v-lambda*v for reference eigenvectors:
3.35950672288886e-16
3.6369535193841e-16
3.63594787451443e-16
```
Notice that reference eigenvectors are much better fit than what Spectra can find. I tried increasing `ncv_`, and it somewhat improved the precision, but after some value the precision got worse again.
Notice also that `tol` here was set to a ridiculously small value (originally it was simply `...epsilon()`, only in later attempts to get higher precision I multiplied by that `1e-300` constant), but still Spectra doesn't appear to be able to converge to a good approximation for eigenvectors, nor to report any failure.

**Comment by yixuan** [comment]
Thanks for the report. Could you also test with ARPACK for reference?

**Comment by 10110111** [comment]
ARPACK gives much better results. Test code is [here](https://pastebin.mozilla.org/9077854), and these are the results:
```
nconv: 3, iterations taken: 5350
Eigenvalues found:
0.493480220054469
0.493480220054446
0.19739208802176
Difference from reference eigenvalues:
1.54321000422897e-14
-7.66053886991358e-15
-1.33504318711175e-14
Standard deviations of H*v-lambda*v for computed eigenvectors:
4.90161722735067e-14
4.32942649601542e-14
1.84667811779593e-14
Standard deviations of H*v-lambda*v for reference eigenvectors:
4.83009936746075e-16
4.96137214842756e-16
5.2848569561464e-16
```

**Comment by yixuan** [comment]
I see. I'll keep in mind and try to fix it.

**Comment by yixuan** [comment]
I have modified the code a little bit in the [develop](https://github.com/yixuan/spectra/tree/develop) branch, which you may want to have a try.

I cannot say that the accuracy has already matched APRACK in this example, but it indeed has improved in my machine. I will continue studying the reason and try to refine the algorithm.

**Comment by EDMIG** [comment]
For large matrix,  suggest to increase subspace-dimension of Lanzcos basis accordingly, and there will be some gains on convergence&accuracy!
 
test on matlab:

A=rand(2000);
d0=eigs(A,6,'LM')

%set subspace dimension of Lanczos basis
opt.p=20;
d1=eigs(A,6,'LM',opt)

opt.p=50;
d2=eigs(A,6,'LM',opt)

output:
警告: 仅 6 请求特征值的 3 已经聚合。 
> In eigs>CallARPACK/processEUPDinfo (line 1111)
  In eigs>CallARPACK (line 1054)
  In eigs (line 113) 

d0 =

   1.0e+03 *

  1.000157484483733 + 0.000000000000000i
 -0.004586383503817 - 0.012269221841457i
 -0.004586383503817 + 0.012269221841457i
                NaN +               NaNi
                NaN +               NaNi
                NaN +               NaNi

警告: 仅 6 请求特征值的 3 已经聚合。 
> In eigs>CallARPACK/processEUPDinfo (line 1111)
  In eigs>CallARPACK (line 1054)
  In eigs (line 113) 

d1 =

   1.0e+03 *

  1.000157484483733 + 0.000000000000000i
 -0.004586383503817 - 0.012269221841457i
 -0.004586383503817 + 0.012269221841457i
                NaN +               NaNi
                NaN +               NaNi
                NaN +               NaNi


d2 =

   1.0e+03 *

  1.000157484483733 + 0.000000000000000i
 -0.004586383503817 - 0.012269221841457i
 -0.004586383503817 + 0.012269221841457i
  0.013030318632097 + 0.000000000000000i
  0.006750326886985 + 0.011113419540869i
  0.006750326886985 - 0.011113419540869i

**Comment by yixuan** [comment]
@10110111 Hi, I want to revisit this problem, but the code files are now invalid. Do you still have a copy? Thanks.

**Comment by 10110111** [comment]
Here are they: [spectra-test.tar.gz](https://github.com/yixuan/spectra/files/4662865/spectra-test.tar.gz).

**Comment by yixuan** [comment]
Thanks! This is very helpful.

**Comment by yixuan** [comment]
I think the precision has been greatly improved in the [1.y.z](https://github.com/yixuan/spectra/tree/1.y.z) branch, which is meant to be the next major version. One big change in this branch is that API has changed ([changelog](https://github.com/yixuan/spectra/commit/d6b33b8961bd90186b96c15f31625b4cb44d0ba6)), but the migration is straightforward. I changed your main function to the following:

```cpp
int main(int argc, char** argv)
{
    const auto H = makeH();
    SparseSymMatProd matrixOperator(H);
    Spectra::SymEigsSolver<Real, SparseSymMatProd> eigs(matrixOperator, 3, 6);

    try
    {
        eigs.init();
        const auto tolerance=std::numeric_limits<Real>::epsilon();
        eigs.compute(Spectra::SortRule::SmallestAlge, 10000, tolerance);
    }
    catch(std::exception& ex)
    {
        std::cerr << "Error: " << ex.what() << "\n";
        return 1;
    }

    if(eigs.info() != Spectra::CompInfo::Successful)
    {
        std::cerr << "Failed to compute eigenvalues.\n";
        return 1;
    }
    std::cout.precision(std::numeric_limits<Real>::digits10);
    const auto evals = eigs.eigenvalues();
    const auto evecs = eigs.eigenvectors();
    std::cout << "Eigenvalues found in "
              << eigs.num_iterations() << " iterations, "
              << eigs.num_operations() << " operations:\n"
              << evals << '\n';

    const std::vector<Real> refEvals{0.49348022005445374,0.49348022005445374,0.197392088021773};
    std::cout << "Difference from reference eigenvalues:\n";
    for(int i=0;i<refEvals.size();++i)
        std::cout << evals(i,0)-refEvals[i] << "\n";

    // Some other code
    return 0;
}
```

Then I got the results as below:
```
Eigenvalues found in 4465 iterations, 10379 operations:
0.493480220054545
0.493480220054499
 0.19739208802177
Difference from reference eigenvalues:
9.12603326241879e-14
4.57411886145564e-14
-2.99760216648792e-15
Standard deviations of H*v-lambda*v for computed eigenvectors:
3.16860150588067e-14
2.15824003060412e-14
1.77922007750191e-14
Standard deviations of H*v-lambda*v for reference eigenvectors:
7.93927242175428e-15
6.29144133912649e-15
8.56253294843173e-15
```

Note that I compiled with `-O3 -mtune=native` instead of `-O3 -march=native`, since the latter may break some floating number operation rules. You may see slightly larger error if you use the latter option.

**Comment by 10110111** [comment]
> since the latter may break some floating number operation rules

It shouldn't: it doesn't enable any of the constituent options of `-ffast-math`, nor any other rule-breaking options. But what it does enable on my system, and what leads to lower precision is `-mavx2`, which, again, shouldn't result in breaking C++ floating-point rules.

Also, I've noticed you've renamed `include/spectra` to `include/Spectra`. Is it intentional as an additional change to the changes in API?

**Comment by yixuan** [comment]
It may change the order of floating number operations. For example, I want to compute
```
x[0] + x[1] + ... + x[7]
```
Normally the loop is
```
r = x[0]
r = r + x[1]
...
r = r + x[7]
```
With AVX2 it is likely to be
```
v = (x[0], ..., x[3])
v = v + (x[4], ..., x[7]) = (x[0] + x[4], x[1] + x[5], x[2] + x[6], x[3] + x[7])
r = v[0] + v[1] + v[2] + v[3]
```

I didn't change `spectra` to `Spectra`. Previously there was no such a folder, and then I added the `Spectra` folder ([changelog](https://github.com/yixuan/spectra/commit/7bd8cedeb671b16cc318272fe6adee05df4aa62c#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR4-R6)).

**Comment by 10110111** [comment]
Well, if such vectorization will happen with AVX2 by default, then it's already broken: SSE2 is on by default on x86_64, and this ISA is capable of exactly the same kind of vectorization, just with shorter vectors. Only x86-32 doesn't assume SSE2 is available.

**Comment by yixuan** [comment]
Yeah, I guess the ordering results in some small random errors. Maybe in another case the precision is higher with vectorization enabled.

**Comment by 10110111** [comment]
At least according to [this](https://stackoverflow.com/a/59646157/673852), reductions are not vectorized by default, unless you use `-fassociative-math` (or enable it via `-ffast-math` etc.).

**Comment by yixuan** [comment]
Oh, this is done by Eigen, not the auto-vectorization by compilers.
